### PR TITLE
[swift] swift prometheus alerts

### PR DESCRIPTION
**Pull request by RohithChandran**

### CODE REVIEW

**Comment by thgrs** [comment]
hi rohit, all the max(x,180) (dito for sum, avg, etc.) have to be converted to max_over_time(x[3m]) for prometheus.

**Comment by thgrs** [comment]
some more observations: i'm not sure if the "by" clause is correct - example: the first alert - swift-account-first-byte-timing in monasca:

max(swift.proxy.firstbyte_timer{status=200,type=account,quantile=0.5}, 180)

and

        "match_by": [
            "service",
            "hostname"
        ],

that should get max_over_time(swift_proxy_firstbyte_timer{status="200",type="account",quantile="0.5"}[3m]) by (instance)

i.e.: 
* swift.proxy.... -> swift_proxy_... - done
* status=200 -> status="200" - done
* max(x,180) -> max_over_time(x[180]) - still missing
* match_by "service", "hostname" -> by (instance) - as hostname comes from instance and service is the same always anyway - still missing

@auhlig does all the above make sense?

**Comment by auhlig** [comment]
The`by (cluster)` is used to group alerts. So that's correct.

**Comment by reimannf** [comment]
Should we use `ALERT ... FOR 3m` or `max_over_time()...[3m]` ?

**Comment by auhlig** [comment]
Ah. Now that I think about it, I might need to reconsider my advice from this morning. The `max()` is already what we want, since it is evaluated `FOR 3m`.

**Comment by auhlig** [comment]
I'll take care of removing that in the alerts comitted this morning

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 1]
```diff
@@ -0,0 +1,184 @@
+ALERT SwiftAccountFirstByteTiming
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Please generalize FirstByteTiming. They are the same for account, container and object.

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  IF max(swift_proxy_firstbyte_timer{status="200",type="account",quantile="0.5"}) by (cluster) > 500.0
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Replace by `max(swift_proxy_firstbyte_timer{status="200",quantile="0.5"}) by (os_cluster, type) > 1000.0`

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "critical"
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: `info` is enough

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    context = "openstack"
+    dashboard = "swift-overview"
+  }
+  ANNOTATIONS {
+    summary = "swift-account-first-byte-timing",
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Please use labels `os_cluster` and `type` in summary

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    description = "Alert to notify if the swift account first-byte-timing reached more than 500 within 3 minutes.",
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Until we have a playbook... `Indicates latency for token validation`

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 6]
```diff
@@ -0,0 +1,184 @@
+ALERT SwiftAccountFirstByteTiming
+  IF max(swift_proxy_firstbyte_timer{status="200",type="account",quantile="0.5"}) by (cluster) > 500.0
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "critical"
+    context = "openstack"
+    dashboard = "swift-overview"
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: `swift-proxy` dashboard. There are vars for cluster and proxy instance for preselection available like ```swift-proxy?var-cluster={{`{{$labels.os_cluster}}`}}&var-proxy={{`{{$labels.instance}}`}}```

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  }
+  ANNOTATIONS {
+    summary = "swift-account-first-byte-timing",
+    description = "Alert to notify if the swift account first-byte-timing reached more than 500 within 3 minutes.",
+  }
+
+ALERT SwiftContainerFirstByteTiming
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: That can go. See above

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    severity = "critical"
+    context = "openstack"
+    dashboard = "swift-overview"
+  }
+  ANNOTATIONS {
+    summary = "swift-container-first-byte-timing",
+    description = "Alert to notify if the swift container first-byte-timing reached more than 500 within 3 minutes.",
+  }
+  
+ALERT SwiftObjectFirstByteTiming
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: That can go. See above

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    summary = "swift-used-space-70",
+    description = "Swift storage usage reached 70%",
+  }
+  
+ALERT SwiftOverLappingObjects
+  IF min(swift_dispersion_container_overlapping_gauge{}) by (cluster) > 0
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "warning"
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: `info` is enough

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    summary = "swift-overlapping-objects",
+    description = "Swift has overlapping objects. Run swift-dispersion-report to get details",
+  }
+  
+ALERT SwiftOverLappingContainers
+  IF min(swift_dispersion_container_overlapping_gauge{}) by (cluster) > 0.0
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "warning"
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: `info` is enough

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    context = "openstack"
+    dashboard = "swift-overview"
+  }
+  ANNOTATIONS {
+    summary = "swift-mismatched-rings",
+    description = "Alarm on mismatched rings",
+  } 
+  
+ALERT SwiftDisksUnmounted
+  IF max(swift_cluster_drives_unmounted_gauge{}) by (cluster) > 0.0
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Please use `sum(swift_cluster_drives_unmounted_gauge) by (storage_ip)`

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "warning"
+    context = "openstack"
+    dashboard = "swift-overview"
+  }
+  ANNOTATIONS {
+    summary = "swift-disks-unmounted",
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: add label `storage_ip` to summary

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    description = "Swift drives not mounted. Run swift-recon --unmounted to get details",
+  } 
+
+ALERT SwiftDiskFailures
+  IF max(swift_cluster_drives_audit_errors_gauge{}) by (cluster) > 0.0
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Please use `sum(swift_cluster_drives_audit_errors_gauge) by (storage_ip)`

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "warning"
+    context = "openstack"
+    dashboard = "swift-overview"
+  }
+  
+  ANNOTATIONS {
+    summary = "swift-disk-failures",
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: add label `storage_ip` to summary

File: system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+ALERT SwiftUsedSpace50
+  IF max(swift_cluster_storage_used_percent_gauge{}) by (cluster) > 0.5
+  FOR 3m
+  LABELS {
+    service = "swift"
+    severity = "info"
+    context = "openstack"
+    dashboard = "swift-overview"
+  }
+
```
Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Please remove that space

Comment (reimannf) [inline system/kube-monitoring/charts/prometheus-frontend/openstack-swift.alerts]: Please order and place all capaictity alerts together

**Pull request merged**